flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/TableScan.java

```diff
@@ -0,0 +1,151 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.table.source;
+
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.table.data.binary.BinaryRowData;
+import org.apache.flink.table.store.file.data.DataFileMeta;
+import org.apache.flink.table.store.file.operation.FileStoreScan;
+import org.apache.flink.table.store.file.predicate.And;
+import org.apache.flink.table.store.file.predicate.CompoundPredicate;
+import org.apache.flink.table.store.file.predicate.LeafPredicate;
+import org.apache.flink.table.store.file.predicate.Predicate;
+import org.apache.flink.table.store.file.predicate.PredicateBuilder;
+import org.apache.flink.table.store.file.schema.Schema;
+import org.apache.flink.table.store.file.utils.FileStorePathFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/** An abstraction layer above {@link FileStoreScan} to provide input split generation. */
+public abstract class TableScan {
+
+    protected final FileStoreScan scan;
+    private final int[] fieldIdxToPartitionIdx;
+    private final FileStorePathFactory pathFactory;
+
+    protected TableScan(FileStoreScan scan, Schema schema, FileStorePathFactory pathFactory) {
+        this.scan = scan;
+        List<String> partitionKeys = schema.partitionKeys();
+        this.fieldIdxToPartitionIdx =
+                schema.fields().stream().mapToInt(f -> partitionKeys.indexOf(f.name())).toArray();
+        this.pathFactory = pathFactory;
+    }
+
+    public TableScan withSnapshot(long snapshotId) {
+        scan.withSnapshot(snapshotId);
+        return this;
+    }
+
+    public TableScan withFilter(Predicate predicate) {
+        List<Predicate> partitionFilters = new ArrayList<>();
+        List<Predicate> nonPartitionFilters = new ArrayList<>();
+        for (Predicate p : PredicateBuilder.splitAnd(predicate)) {
+            Optional<Predicate> mapped = mapToPartitionFilter(p);
+            if (mapped.isPresent()) {
+                partitionFilters.add(mapped.get());
+            } else {
+                nonPartitionFilters.add(p);
+            }
+        }
+
+        scan.withPartitionFilter(new CompoundPredicate(And.INSTANCE, partitionFilters));
+        withNonPartitionFilter(new CompoundPredicate(And.INSTANCE, nonPartitionFilters));
+        return this;
+    }
+
+    public Plan plan() {
+        FileStoreScan.Plan plan = scan.plan();
+        List<Split> splits = new ArrayList<>();
+        for (Map.Entry<BinaryRowData, Map<Integer, List<DataFileMeta>>> entryWithPartition :
+                plan.groupByPartFiles().entrySet()) {
+            BinaryRowData partition = entryWithPartition.getKey();
+            for (Map.Entry<Integer, List<DataFileMeta>> entryWithBucket :
```

Comment:
These logical should in a `SplitGenerator`, it can be improved.
See https://github.com/apache/flink-table-store/pull/131